### ISSUE TITLE
Missing getServerSnapshot on React 18 and Next 12

### ISSUE THREAD

**Issue by padenj** [apollographql/apollo-client#9623]
**Intended outcome:**
Adding an Apollo Client to a Next JS application with React 18 should execute the query and load the data.  However, the error message below prevents the page from loading.  

**Actual outcome:**
error - Error: Missing getServerSnapshot, which is required for server-rendered content. Will revert to client rendering.
Error: Missing getServerSnapshot, which is required for server-rendered content. Will revert to client rendering.
    at Object.useSyncExternalStore (node_modules/react-dom/cjs/react-dom-server.browser.development.js:5051:11)
    at Object.useSyncExternalStore (/node_modules/react/cjs/react.development.js:1671:21)
    at InternalState.useQuery (/node_modules/@apollo/client/react/hooks/hooks.cjs:70:31)
    at useQuery (/node_modules/@apollo/client/react/hooks/hooks.cjs:28:69)

**How to reproduce the issue:**
* Use `create-next-app` to create a new NextJS application with React 18 and Next 12.  
* Setup an Apollo Server endpoint similar to the steps discussed in this blog post: https://www.prisma.io/blog/fullstack-nextjs-graphql-prisma-2-fwpc6ds155
* Install @apollo/client@beta (currently 3.6.0-rc.1) and configure a query similar to the above blog post.  The beta is currently required to install Apollo Client with React 18.
* Executing the client query results in the error message "Error: Missing getServerSnapshot, which is required for server-rendered content. Will revert to client rendering."

A similar issue was identified and fixed in another package, Zustand.   https://github.com/pmndrs/zustand/commit/773e2e2919d5367a41f27adaa63fe2ab2e572d07

**Versions**
 System:
    OS: macOS 12.3.1
  Binaries:
    Node: 17.1.0 - ~/.nvm/versions/node/v17.1.0/bin/node
    Yarn: 1.22.17 - ~/.nvm/versions/node/v17.1.0/bin/yarn
    npm: 8.1.2 - ~/.nvm/versions/node/v17.1.0/bin/npm
  Browsers:
    Chrome: 100.0.4896.127
    Firefox: 99.0.1
    Safari: 15.4
  npmPackages:
    @apollo/client: ^3.6.0-rc.1 => 3.6.0-rc.1 
    apollo-server-micro: ^3.5.0 => 3.5.0

**Comment by Juan-seb** [comment]
I believe this issue is related to the migration of react 17 to react 18, i have the same problem and fix this only downgraded react (version 18 to 17)

**Comment by w3bdesign** [comment]
@padenj 

Can you try updating to the latest v3.6 beta with `npm i @apollo/client@beta` ?

**Comment by yura3d** [comment]
3.6.0 has just released (available now as `@apollo/client@next` in npm) and SSR is still broken:
```
Error: Missing getServerSnapshot, which is required for server-rendered content. Will revert to client rendering.
    at Object.useSyncExternalStore (***\node_modules\react-dom\cjs\react-dom-server.browser.development.js:5051:11)
    at Object.useSyncExternalStore (***\node_modules\react\cjs\react.development.js:1671:21)
    at InternalState.useQuery (***\node_modules\@apollo\client\react\hooks\hooks.cjs:70:31)
    at useQuery (***\node_modules\@apollo\client\react\hooks\hooks.cjs:28:69)
    at *** (webpack-internal:///./***)
    at renderWithHooks (***\node_modules\react-dom\cjs\react-dom-server.browser.development.js:5448:16)
    at renderIndeterminateComponent (***\node_modules\react-dom\cjs\react-dom-server.browser.development.js:5521:15)
    at renderElement (***\node_modules\react-dom\cjs\react-dom-server.browser.development.js:5736:7)
    at renderNodeDestructive (***\node_modules\react-dom\cjs\react-dom-server.browser.development.js:5875:11)
    at renderContextProvider (***\node_modules\react-dom\cjs\react-dom-server.browser.development.js:5710:3)
error - Error: Missing getServerSnapshot, which is required for server-rendered content. Will revert to client rendering.
```

Windows 11 x64 / Debian 10 x64
Node.js 17.8.0
react 18.0.0
react-dom 18.0.0
next 12.1.5
@apollo/client 3.6.0
graphql 15.8.0

**Comment by bigint** [comment]
This is happening also to me in with CSR app 😕

**Comment by yura3d** [comment]
CSR works in CRA apps, but doesn't work in Next.js apps. Next.js runs hooks on both sides (server and client), and exactly running on a server triggers the error.

**Comment by bluematter** [comment]
I temporarily solved it by doing this

```
import React, { useEffect, useState } from "react";
import useUser from "../../src/lib/hooks/useUser";
import Admin from "../../src/containers/Dashboard/Admin";

const AdminPage = ({ ssrMode }) => {
  const [isClient, setClient] = useState<boolean>(false);
  const { user } = useUser({});

  useEffect(() => {
    setClient(true);
  }, []);

  return isClient ? <Admin user={user} ssrMode={ssrMode} /> : null;
};

export default AdminPage;
```

**Comment by HT-Moh** [comment]
I am facing the same issue: 

![image](https://user-images.githubusercontent.com/34759569/165744421-d6632727-548c-4fa0-8216-84aaf0ece55a.png)

```
"@apollo/client
"@next-auth/prisma-adapter": "^1.0.3",
"@prisma/client": "^3.13.0",
"next": "^12.1.5",
"react": "^18.1.0",
"react-dom": "^18.1.0",
 "@types/react": "^18.0.8",
 "prisma": "^3.13.0",
```

**Comment by pin4eva** [comment]
The error is from react 18. To solve this problem simply downgrade your react version

**Comment by markedwards** [comment]
Confirmed, this appears to be a bug in @apollo/client >= 3.6.0.  Downgrading to <3.6.0 fixes it.

**Comment by kkomelin** [comment]
Same here. Setting ` "@apollo/client": "3.5.10"` has fixed the issue.

**Comment by MYKEU** [comment]
Experience the same after updating to `3.6.1`

**Comment by tranlehaiquan** [comment]
Same problem after update 3.6.1
![image](https://user-images.githubusercontent.com/17347993/166143519-01cc0fbc-aaec-4c3b-8fef-d00cb355c502.png)

**Comment by hungphongbk** [comment]
I've created a small bugfix PR for this #9652 . Just waiting for approval

**Comment by tranlehaiquan** [comment]
@hungphongbk ,
That great. I have never known about `useSyncExternalStore`. Thank you.

**Comment by mohammed-gehad** [comment]
this bug happens when a page has a **server side render** and a useQuery hook, the hook fires at server render which causes the problem, **make sure it fires only when has rendered on client side**, you can make that with the help of use effect hook.
here is a custom component to fix this issue. 

`// components/ClientOnly.js

import { useEffect, useState } from "react";

export default function ClientOnly({ children, ...delegated }) {
  const [hasMounted, setHasMounted] = useState(false);

  useEffect(() => {
    setHasMounted(true);
  }, []);

  if (!hasMounted) {
    return null;
  }

  return <div {...delegated}>{children}</div>;
}`

read this post for more info https://www.joshwcomeau.com/react/the-perils-of-rehydration/

**Comment by elrumordelaluz** [comment]
for me is solved in `@apollo/client@3.6.2`

**Comment by benjamn** [comment]
Thanks everyone!
